### DISCUSSION TITLE
RFC: Should we make all redirects on Dub 302 by default?

### DISCUSSION THREAD

**Discussion by steven-tey** [dubinc/dub#659] [Polls]
Update (Mar 9, 2024): This has been shipped: https://dub.co/changelog/redirect-status-codes

Edit: Have been doing a bit more research and it looks like 302 redirect is the best option here since unlike 301, it doesn’t cache the destination URL in the user’s browser (in case you made a typo/need to update the URL). It’s also ideal for SEO as per [Google themselves](https://youtu.be/6LkKSn3CiP4?si=2bRFePmYn9SBOBhk).

Hey everyone! Building off of #182 – we've gotten a bunch of requests to support ~~301~~ 302 redirects recently, and since we haven't really heard anyone specifically request 307 redirects, we're thinking of changing all redirects on Dub to be ~~301~~ 302 by default (for better SEO). Here's a [1-line PR](https://github.com/dubinc/dub/pull/664) of the change.

Alternatively, we could also set 301 as the default for [root domain redirects](https://dub.co/help/article/how-to-redirect-root-domain) and 302 for all other links since you're less likely to update your root domain redirect.

In the future, we might add more granular options for each link (301, 307, 308), but for now, ~~301~~ 302 will be the default redirect behavior.

What do you guys think?

**Comment by zlwaterfield** [comment]
This is a good move and probably better follows the practices of link shorteners. 

The one edge case to be aware of is if the link owner changes the destination URL after a person has already visited the link, the browser will still visit the old destination URL since it catches the URL on a 301. I think this is a pretty large edge case, though. 

You could offer up the ability to choose and default to 301. Basically, if the user plans to change the destination URL, then they could use a 307.

**Reply by steven-tey** [reply]
that's a very good point, thanks for weighing in Zach!

**Comment by Illyism** [comment]
Defaulting to 301 redirects for all redirects in Dub may limit flexibility for users.

- **301 redirects** are seen as **permanent** by search engines and browsers.
- If a user changes the destination URL later, it may cause confusion for search engines.
- **302 redirects** offer more flexibility as they are **temporary**.
- **301 redirects** are **cached** in the browser, which can prevent users from seeing changes if they visit the same URL again and the destination has changed in Dub.
- Ideally, there should be a simple explanation to explain the differences between 301s and 302s.
- Maybe this would be good to have in the options? But this really depends on the use cases for each person. I like having the ability to change my destination URL later (if I forget utm tags, to fix issues, ...)
∙ Make sure to clearly communicate the implications of using 301 redirects to users.

In short, relying solely on 301 redirects limits the ability to modify destination URLs safely.

**Comment by versecafe** [comment]
Noting the two comments above, having it configureable by the user may be better then always doing 301 redirects, perhaps giving the user a switch to pick from either `Constant` (301) or `Mutable` (302/307), the phrasing could be changed around but leaving it up to users and then making it impossible for them to change the destination if `Constant` is selected.
